cmd/root.go

```diff
@@ -15,55 +15,61 @@
 package cmd
 
 import (
+	"context"
 	"fmt"
-	"github.com/fairwindsops/astro/pkg/metrics"
-	"github.com/prometheus/client_golang/prometheus/promhttp"
 	"net/http"
 	"os"
 	"os/signal"
 	"strings"
 	"syscall"
+	"time"
 
 	"github.com/fairwindsops/astro/pkg/controller"
+	"github.com/fairwindsops/astro/pkg/kube"
+	"github.com/fairwindsops/astro/pkg/metrics"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
 	log "github.com/sirupsen/logrus"
 	"github.com/spf13/cobra"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
 )
 
-var logLevels = map[string]log.Level{
-	"panic": log.PanicLevel,
-	"fatal": log.FatalLevel,
-	"error": log.ErrorLevel,
-	"warn":  log.WarnLevel,
-	"info":  log.InfoLevel,
-	"debug": log.DebugLevel,
-	"trace": log.TraceLevel,
-}
-
-var rootCmd = &cobra.Command{
-	Use:   "astro",
-	Short: "Kubernetes datadog monitor manager",
-	Long:  "A kubernetes agent that manages datadog monitors.",
-	Run:   run,
-}
+var (
+	logLevels = map[string]log.Level{
+		"panic": log.PanicLevel,
+		"fatal": log.FatalLevel,
+		"error": log.ErrorLevel,
+		"warn":  log.WarnLevel,
+		"info":  log.InfoLevel,
+		"debug": log.DebugLevel,
+		"trace": log.TraceLevel,
+	}
+	rootCmd = &cobra.Command{
+		Use:   "astro",
+		Short: "Kubernetes datadog monitor manager",
+		Long:  "A kubernetes agent that manages datadog monitors.",
+		Run:   leaderElection,
+	}
+	logLevel    string
+	metricsPort string
+	namespace   string
+)
 
-var logLevel string
-var metricsPort string
+const (
+	defaultLeaseDuration = 15 * time.Second
+	defaultRenewDeadline = 10 * time.Second
+	defaultRetryPeriod   = 2 * time.Second
+)
 
 func init() {
 	rootCmd.PersistentFlags().StringVarP(&logLevel, "level", "l", "info", "Log level setting. Default is INFO. Should be one of PANIC, FATAL, ERROR, WARN, INFO, DEBUG, or TRACE")
 	rootCmd.PersistentFlags().StringVarP(&metricsPort, "metrics-port", "p", ":8080", "The address to serve prometheus metrics.")
+	rootCmd.PersistentFlags().StringVar(&namespace, "namespace", "kube-system", "The namespace where astro is running")
 }
-
-func run(cmd *cobra.Command, args []string) {
```

Comment:
One thing I'll call out is that I removed this line so logging doesn't output the full file path to where a log is located within the project. I could envision adding this back with a debug flag or if log-level is set to debug for development purposes, but I find it to be way too noisy.

I'd be happy to add it back with those debug flag stipulations if folks think it provides value.